pallets/liquid-staking-two/src/protocol.rs

```diff
@@ -0,0 +1,140 @@
+                user_ledger_per_era.total_stake_amount = user_ledger_per_era
+                    .total_stake_amount
+                    .checked_add(amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPoolByEra::<T>::try_mutate(
+            &Self::current_era(),
+            |pool_ledger_per_era| -> DispatchResult {
+                pool_ledger_per_era.total_stake_amount = pool_ledger_per_era
+                    .total_stake_amount
+                    .checked_add(amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+        Ok(().into())
+    }
+
+    // After confirmed unbond on relaychain,
+    // and then burn/withdraw xKSM.
+    // before update exchangerate (record_reward)
+    fn unstake(who: &T::AccountId, amount: BalanceOf<T>) -> Result<BalanceOf<T>, DispatchError> {
+        // can not burn directly because we have match mechanism
+        T::Currency::transfer(T::LiquidCurrency::get(), who, &Self::account_id(), amount)?;
+
+        let exchange_rate = ExchangeRate::<T>::get();
+        let asset_amount = exchange_rate
+            .checked_mul_int(amount)
+            .ok_or(Error::<T>::InvalidExchangeRate)?;
+
+        MatchingQueueByUser::<T>::try_mutate(
+            who,
+            &Self::current_era(),
+            |user_ledger_per_era| -> DispatchResult {
+                user_ledger_per_era.total_unstake_amount = user_ledger_per_era
+                    .total_unstake_amount
+                    .checked_add(asset_amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+
+        MatchingPoolByEra::<T>::try_mutate(
+            &Self::current_era(),
+            |pool_ledger_per_era| -> DispatchResult {
+                pool_ledger_per_era.total_unstake_amount = pool_ledger_per_era
+                    .total_unstake_amount
+                    .checked_add(asset_amount)
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                Ok(())
+            },
+        )?;
+        Ok(asset_amount)
+    }
+
+    fn claim(who: &T::AccountId) -> DispatchResult {
+        let mut withdrawable_stake_amount = 0u128;
+        let mut withdrawable_unstake_amount = 0u128;
+        let mut remove_record_from_user_queue = Vec::<EraIndex>::new();
+        MatchingQueueByUser::<T>::iter_prefix(who).for_each(|(era_index, user_ledger_per_era)| {
+            Self::accumulate_claim_by_era(
+                who,
+                era_index,
+                user_ledger_per_era,
+                &mut withdrawable_stake_amount,
+                &mut withdrawable_unstake_amount,
+                &mut remove_record_from_user_queue,
+            );
+        });
+
+        // remove finished records from MatchingQueue
+        if remove_record_from_user_queue.len() > 0 {
```

Comment:
`!is_empty` should be preferred

Reply:
what's the different?

Reply:
we don't have hard coded number in the code, you can check loans pallet

Reply:
https://doc.rust-lang.org/stable/src/alloc/vec/mod.rs.html#1842

Reply:
yes but `!is_empty` is easier to read